ts/src/main/scala/ch/epfl/ts/engine/rules/FxMarketRulesWrapper.scala

```diff
@@ -0,0 +1,90 @@
+package ch.epfl.ts.engine.rules
+
+import ch.epfl.ts.engine.{OrderBook, ForexMarketRules}
+import ch.epfl.ts.data._
+import ch.epfl.ts.data.Currency.Currency
+import scala.collection.mutable
+import scala.Some
+import ch.epfl.ts.data.MarketBidOrder
+import ch.epfl.ts.data.LimitBidOrder
+import scala.Some
+import ch.epfl.ts.data.MarketAskOrder
+import ch.epfl.ts.data.LimitAskOrder
+
+/**
+ * Wrapper around ForexMarketRules, to have ALL the information needed to proceed the order in one method.
+ * Created by sygi on 09.05.15.
+ */
+class FxMarketRulesWrapper extends MarketRulesWrapper(new ForexMarketRules()) {
+  val rules = new ForexMarketRules()
+  override def processOrder(o: Order, marketId: Long,
+                            book: OrderBook, tradingPrices: mutable.HashMap[(Currency, Currency), (Double, Double)],
+                            send: Streamable => Unit): Unit =
+  o match {
+    case limitBid: LimitBidOrder =>
+      book.bids.insert(limitBid)
+      checkPendingOrders(marketId, book, tradingPrices, send)
+
+    case limitAsk: LimitAskOrder =>
+      book.asks.insert(limitAsk)
+      checkPendingOrders(marketId, book, tradingPrices, send)
+
+    case marketBid: MarketBidOrder =>
+      println("FMRW : received a bidOrder")
+      tradingPrices.get((marketBid.withC, marketBid.whatC)) match {
+        // We buy at current ask price
+        case Some(t) => rules.matchingFunction(marketId, marketBid, book.bids, book.asks, send, t._2)
+        case None    =>
+      }
+    case marketAsk: MarketAskOrder =>
+      println("FMRW : received an askOrder")
+      tradingPrices.get((marketAsk.withC, marketAsk.whatC)) match {
+        // We sell at current bid price
+        case Some(t) => rules.matchingFunction(marketId, marketAsk, book.bids, book.asks, send, t._1)
+        //TODO(sygi): does it actually work at all, when there are multiple currencies?
```

Comment:
I thought about it and it actually don't work, as there is only one instance of `OrderBook`, so one cannot take the top bid/ask from the given currency.